encoding/protobuf/reflection/reflection.go

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package reflection
+
+// ServerReflectionInfo encapsulates the information about a service required
+// for exposing this service on the gRPC reflection protocol.
+// See https://github.com/grpc/grpc/blob/master/doc/server-reflection.md
+type ServerReflectionInfo struct {
+
+	// The fully qualified service name
+	ServiceName string
+
+	// RegisteredFileName is the identifer with hich the filedescriptor
+	// associated with this service is registered using `proto.RegisterFile()`
+	// during init. It can be used to get a compressed filedescriptor using
+	// `proto.FileDescriptor()`.
+	RegisteredFileName string
```

Comment:
For further reference, this is what this relies on: https://github.com/golang/protobuf/blob/master/proto/properties.go#L535

That is accessing from the global map `protoFiles ` https://github.com/golang/protobuf/blob/master/proto/properties.go#L525, which there is a different global map in gogo/protobuf `protoFiles ` https://github.com/gogo/protobuf/blob/master/proto/properties.go#L590

Reply:
Attaching the `FileDescriptorProto` directly instead of using these names was what I had originally, but that also ran into some problems:

- The variable name identifying the descriptor in the `.pb.go` is not stable across gogo/golang proto. The only way to reliably get that varname is by making the yarpc-go generator a plugin/option of the general `protoc-gen-{go,gogo}`. I was looking up why to comment here and it also seems they chagned it again two days ago https://github.com/golang/protobuf/pull/696, so I think that signifies more that only through tight coupling of the yarpc generator and the other generator we could achieve that.
 - Generating the filedescriptor again in `.pb.yarpc.go` could get around this but also has problems: as the gogo and golang descriptor are not the same, the contents will not be the same. Same problem here about yarpc should not assume which one you want.
- Dependencies. If you have a dependency on another file, you need to access the filedescriptor for that file. I don't believe the serialised filedescriptors contain dependencies, only filenames of their dependencies. You'd still need to access them somehow. The options I see is through some sort of registrty, either compatible with golang/gogo (which results in the current problem) or local to yarpc, but that means we need to change the codegen to always generate filedescriptors to register (including if there are no services, but that would still require the decision about which generator type format to serialize in the point above).

I don't see any good ways forward here.

Reply:
How are the descriptors different? They are a language (and plugin) independent construct.

I would take the descriptor on the CodeGeneratorRequest and output it serialized into .pb.yarpc.go yourself. I would need more context as to why they are different between golang and gogo, the input file descriptor is the same to both.

Regardless, the proto.FileDescriptor option isn’t stable

Reply:
Also you could just require users to call —include-imports with protoc to use yarpc somehow.

Reply:
> Also you could just require users to call —include-imports with protoc to use yarpc somehow.

`--include_imports` is a `--descriptor_set_out` option, I don't see how that helps? Do you mean replicating the semantics where for every service we don't just serialize the filedescripotor but also the dependencies in a variable during codegen?

Reply:
We should talk offline, but straight up I’m telling you there have been internet-breaking issues (well that’s a little hyperbolic but) due to protoc.FileDescriptor. Yarpc does not depend on globals and for good reason. I’m happy to advise, but using —include_imports with protoc and storing the FileDescriptorSet solves the problem, and regardless a global registry in a variable is not a good long term solution (which is what proto.FileDescriptor is)

Reply:
Ah yea I’m getting things mixed up. Yea I’m not entirely sure how you deal with the dependencies, I don’t know how reflection is handled in other languages. Even if you just do proto.FileDescriptor in a non-global and yarpc-specific way though, that’s better than using proto.FileDescriptor.

Reply:
It turned out that the filedescriptors are actually slightly different, as a result of the extensions in [`gogo.proto`](https://github.com/gogo/protobuf/blob/3cefc55c28cff46426cbe66a7335e37bbf451cb9/gogoproto/gogo.proto):
```
// filedescriptor by gogo
name:"encoding/protobuf/protoc-gen-yarpc-go/internal/testing/testing_no_service.proto" package:"uber.yarpc.encoding.protobuf.protocgenyarpcgo.internal.testing" message_type:<name:"Foo" field:<name:"id" number:1 label:LABEL_OPTIONAL type:TYPE_STRING json_name:"id" > > options:<go_package:"testing" 63002:0 63003:0 63006:1 63008:1 63013:1 63017:1 63018:1 63020:1 63021:0 63022:1 63026:0 > syntax:"proto3" 

// filedescriptor by from yarpc-go
name:"encoding/protobuf/protoc-gen-yarpc-go/internal/testing/testing_no_service.proto" package:"uber.yarpc.encoding.protobuf.protocgenyarpcgo.internal.testing" message_type:<name:"Foo" field:<name:"id" number:1 label:LABEL_OPTIONAL type:TYPE_STRING json_name:"id" > > options:<go_package:"testing" > syntax:"proto3" 
```

I think this isn't a significant difference for the reflection api. Do you think this could become a problem?